packages/loopring_v3/contracts/impl/libexchange/ExchangeBlocks.sol

```diff
@@ -374,50 +263,237 @@ library ExchangeBlocks
                 require(inputEndingHash == endingHash, "INVALID_ENDING_HASH");
                 numWithdrawalRequestsCommitted += uint32(count);
             }
-        } else if (
-            blockType != ExchangeData.BlockType.OFFCHAIN_WITHDRAWAL &&
-            blockType != ExchangeData.BlockType.ORDER_CANCELLATION &&
-            blockType != ExchangeData.BlockType.TRANSFER) {
+        } else if (_block.blockType == ExchangeData.BlockType.TRANSFER) {
+            validateConditionalTransfers(
+                S,
+                _block.data,
+                _block.auxiliaryData
+            );
+        } else if (_block.blockType != ExchangeData.BlockType.OFFCHAIN_WITHDRAWAL) {
             revert("UNSUPPORTED_BLOCK_TYPE");
         }
 
-        // Hash all the public data to a single value which is used as the input for the circuit
-        bytes32 publicDataHash = data.fastSHA256();
-
-        // Store the approved withdrawal data onchain
-        bytes memory withdrawals = new bytes(0);
-        if (blockType == ExchangeData.BlockType.ONCHAIN_WITHDRAWAL ||
-            blockType == ExchangeData.BlockType.OFFCHAIN_WITHDRAWAL) {
+        // Process withdrawals
+        if (_block.blockType == ExchangeData.BlockType.ONCHAIN_WITHDRAWAL ||
+            _block.blockType == ExchangeData.BlockType.OFFCHAIN_WITHDRAWAL) {
             uint start = 4 + 32 + 32;
-            if (blockType == ExchangeData.BlockType.ONCHAIN_WITHDRAWAL) {
+            if (_block.blockType == ExchangeData.BlockType.ONCHAIN_WITHDRAWAL) {
                 start += 32 + 32 + 4 + 4;
             }
-            uint length = 7 * blockSize;
+            uint length = 7 * _block.blockSize;
+            bytes memory withdrawals = new bytes(0);
             assembly {
                 withdrawals := add(data, start)
                 mstore(withdrawals, length)
             }
+            S.distributeWithdrawals(withdrawals);
         }
 
-        // Create a new block with the updated merkle roots
-        ExchangeData.Block memory newBlock = ExchangeData.Block(
-            merkleRootAfter,
-            publicDataHash,
-            ExchangeData.BlockState.COMMITTED,
-            blockType,
-            blockSize,
-            blockVersion,
-            uint32(now),
+        // Emit an event
+        emit BlockSubmitted(S.numBlocksSubmitted, publicDataHash);
+
+        // Transfer the block fee to the fee recipient (for block types with fees paid onchain)
+        withdrawBlockFee(
```

Comment:
Both, depending on the request type. No change was made in v3.5 for this except that the onchain fee can now be paid immediately onchain when the block is submitted because blocks can't be reverted. In v3.0 we had `withdrawBlockFee` that could only withdraw the block fee on finalized blocks.

So for deposits and onchain withdrawals -> onchain block fees in ETH paid by users when doing the request, paid to `feeRecipient` onchain when the request was included in a block.
For all other types of requests only offchain fee payments to the operator's account.